### PR TITLE
Solvers solution pre/post interactions

### LINKED ISSUE #2636: feat: JIT orders pre-interaction support
Excerpt from https://github.com/cowprotocol/services/issues/2607

### PR DESCRIPTION
**Pull request by squadgazzz**
# Description
Adds the ability for solvers to provide pre and post-interactions for a solution.

## How to test
TBD

## Related Issues
Fixes #2636

## Further implementation
1. The same should be done for `gnosis/solvers`
2. Add pre/post-interactions to the Quote response (https://github.com/cowprotocol/services/issues/2680)

### CODE REVIEW

File: crates/driver/src/infra/solver/dto/solution.rs [hunk 1]
```diff
@@ -221,7 +227,11 @@ pub struct Solution {
     #[serde_as(as = "HashMap<_, serialize::U256>")]
     prices: HashMap<eth::H160, eth::U256>,
     trades: Vec<Trade>,
+    #[serde(default)]
+    pre_interactions: Vec<Interaction>,
```
Comment (fleupold) [inline crates/driver/src/infra/solver/dto/solution.rs:231]: I think this is the wrong type. It doesn't make sense for solvers to specify Liquidity interactions as pre interactions. It also doesn't make sense for those to be internalizable nor specify input/output amounts. I think they should simply be 

```
struct {
  target: eth::H160,
  value: eth::U256,
  call_data: Vec<u8>,
}
```

**Review by fleupold — CHANGES_REQUESTED** [review]

File: crates/driver/src/domain/competition/solution/encoding.rs [hunk 2]
```diff
@@ -196,6 +185,33 @@ pub fn tx(
...
+        if matches!(internalization, settlement::Internalization::Enable)
+            && interaction.internalize()
+        {
+            continue;
+        }
+
+        encoded.push(match interaction {
+            competition::solution::Interaction::Custom(interaction) => eth::Interaction {
+                value: interaction.value,
+                target: interaction.target.0.into(),
```
Comment (Mateo-mro) [inline crates/driver/src/domain/competition/solution/encoding.rs:204]: nit

```suggestion
                target: interaction.target.into(),
```

**Review by Mateo-mro — APPROVED** [review]

File: crates/driver/src/infra/solver/dto/solution.rs (hunk 1 above)
Comment (squadgazzz) [reply crates/driver/src/infra/solver/dto/solution.rs]: Switched to `InteractionData`

**Review by MartinquaXD — APPROVED** [review]

**Review by fleupold — APPROVED** [review]
We should note somewhere that this isn't implemented for the boundary settlement encoding (which is fine, I'm hoping to complete the transition by next week)

File: crates/solvers/src/domain/solution.rs [hunk 3]
```diff
@@ -357,6 +361,15 @@ pub struct CustomInteraction {
     pub allowances: Vec<Allowance>,
 }
 
+/// An arbitrary ethereum interaction that is required for the settlement
+/// execution.
+#[derive(Debug)]
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/solvers/src/domain/solution.rs:366]: Didn't get it. To reuse something from the driver crate in what sense?

[the remark this answers is not in the document]
Comment (fleupold) [reply crates/solvers/src/domain/solution.rs:366]: You are redifining a struct in the domain which already exists in the domain. Now, we might not want to overload structs that happen to have the same fields (generally defining new structs for different semantics is good), but here this is similar to re-defining `TokenAmount` (which arguably is never going to change).

File: crates/solvers/src/domain/solution.rs [hunk 4]
```diff
@@ -22,7 +22,9 @@ pub struct Solution {
     pub id: Id,
     pub prices: ClearingPrices,
     pub trades: Vec<Trade>,
+    pub pre_interactions: Vec<InteractionData>,
     pub interactions: Vec<Interaction>,
+    pub post_interactions: Vec<InteractionData>,
```
Comment (fleupold) [inline crates/solvers/src/domain/solution.rs:27]: It would lead to this:

```suggestion
    pub post_interactions: Vec<eth::Interaction>,
```

Comment (squadgazzz) [reply crates/solvers/src/domain/solution.rs:27]: But `eth::Interaction` is part of the driver crate, and the `solver` crate should not depend on the driver, right? Do you want me to extract this type into a shared crate, like `model`?

Comment (fleupold) [reply crates/solvers/src/domain/solution.rs:27]: Oh, you are completely right (I missed that this was a different create). We could still add this type in the eth module of the solvers crate but not super important.

**Pull request merged**